service/frontend/workflow_handler.go

```diff
@@ -3830,26 +3842,66 @@ func (wh *WorkflowHandler) validateBuildIdCompatibilityUpdate(
 }
 
 func (wh *WorkflowHandler) historyArchived(ctx context.Context, request *workflowservice.GetWorkflowExecutionHistoryRequest, namespaceID namespace.ID) bool {
-	if request.GetExecution() == nil || request.GetExecution().GetRunId() == "" {
+	return wh.workflowExecutionArchived(ctx, request.Execution, namespaceID)
+}
+
+func (wh *WorkflowHandler) workflowExecutionArchived(ctx context.Context, execution *commonpb.WorkflowExecution, namespaceID namespace.ID) bool {
+	if execution == nil || execution.GetRunId() == "" {
 		return false
 	}
 	getMutableStateRequest := &historyservice.GetMutableStateRequest{
 		NamespaceId: namespaceID.String(),
-		Execution:   request.Execution,
+		Execution:   execution,
 	}
 	_, err := wh.historyClient.GetMutableState(ctx, getMutableStateRequest)
 	if err == nil {
 		return false
 	}
 	switch err.(type) {
 	case *serviceerror.NotFound:
-		// the only case in which history is assumed to be archived is if getting mutable state returns entity not found error
+		// the only case in which workflow or workflow's event history is assumed to be archived is if getting mutable state returns entity not found error
 		return true
 	}
 
 	return false
 }
 
+func (wh *WorkflowHandler) describeArchivedWorkflowExecution(ctx context.Context, request *workflowservice.DescribeWorkflowExecutionRequest) (*workflowservice.DescribeWorkflowExecutionResponse, error) {
```

Comment:
I don't think archived workflow can be described based on visibility records. Fields like pending activities or pending children can't be populated by vis records. 

If we really want to do it, the right way should be rebuild workflow mutable state from history events, then describe the rebuilt mutable states so that all fields in the describe response can be populated. 

Alternatively, a separate API can be created for describing archived workflow and defined a new response that returns only limited information. 

cc @yiminc